lib/debug/debugger_tooltip.dart

```diff
@@ -0,0 +1,200 @@
+    // backward -> ']' expression | number '['
+    index.set(trim(']') & (expression | digit()) & trim('['));
+    // ref :: identifier [ index ]
+    // backward -> [ index ] identifier
+    ref.set(index.optional() & tagged(id));
+
+    id.set(((letter() | char('_')) & (letter() | digit() | char('_')).star())
+            .flatten());
+
+    reverseContextParser = expression;
+  }
+
+  // Put it back in flowing order.
+  dynamic reverseParse(dynamic value) {
+    if (value is String) {
+      return reverseString(value);
+    } else if (value is List) {
+      return value.reversed.map((v) => reverseParse(v)).toList();
+    } else {
+      return value;
+    }
+  }
+
+  String reverseString(String input) =>
+      new String.fromCharCodes(input.codeUnits.reversed);
+
+  // This is how many characters back at most we go to get the full context.
+  static const backwardBuffer = 160;
+
+  Future check(TooltipElement tooltip) async {
+    HoverInformation info = tooltip.info;
+    TextEditor editor = tooltip.editor;
+
+    int startOffset = math.max(0, info.offset - backwardBuffer);
+    int endOffset = info.offset + info.length;
+    String input = reverseString(editor.getTextInBufferRange(new Range.fromPoints(
+        editor.getBuffer().positionForCharacterIndex(startOffset),
+        editor.getBuffer().positionForCharacterIndex(endOffset))));
+
+    this.endOffset.value = endOffset;
+
+    for (var connection in debugManager.connections) {
+      if (!connection.isAlive) continue;
+
+      DebugVariable result;
+      try {
+        // Parse in reverse from cursor, then let the evaluator unparse it
+        // adding custom context information if needed.
+        var expression = reverseParse(reverseContextParser.parse(input).value);
+        // Let actual debugger to the eval.
+        result = await connection.eval(new DebugExpression(editor.getPath(), expression));
+        if (result == null) return;
+      } catch (e) {
+        _logger.warning(e);
+        break;
+      }
+
+      // If we have anything we create the MTree to display it in the tooltip.
+      MTree<DebugVariable> row = new MTree(new LocalTreeModel(), ExecutionTab.renderVariable)
+          ..flex()
+          ..toggleClass('has-debugger-data');
+
+      row.selectedItem.onChanged.listen((variable) async {
+        if (variable == null) return;
+        await variable.value.invokeToString();
+        await row.updateItem(variable);
+      });
+
+      await row.update([result]);
+
+      String clazz = 'debugger-data';
+      // TODO: this might change after eval of a getter
+      if (!result?.value?.isPrimitive ?? false) clazz += ' expandable';
+      tooltip.expand(div(c: clazz)..add(row));
+    }
+  }
+
+  void dispose() {}
+}
+
+class TooltipEvaluator {
+  final DebugExpression expression;
+
+  TooltipEvaluator(this.expression);
+
+  Future eval() async => visitExpression(expression.expression);
+
+  Future<String> visitExpression(expression) async {
```

Comment:
You may want to explicitly use the `dynamic` type annotation, in order to indicate to readers that the dynamic is intended.

Reply:
Done.